scipy/signal/signaltools.py

```diff
@@ -2920,35 +2937,44 @@ def decimate(x, q, n=None, ftype='iir', axis=-1, zero_phase=None):
     if ftype == 'fir':
         if n is None:
             n = 30
-        system = lti(firwin(n + 1, 1. / q, window='hamming'), 1.)
-
+        system = lti(firwin(n+1, 1. / q, window='hamming'), 1.)
     elif ftype == 'iir':
         if n is None:
             n = 8
         system = lti(*cheby1(n, 0.05, 0.8 / q))
     elif isinstance(ftype, lti):
         system = ftype
+        n = np.max((system.num.size, system.den.size)) - 1
     else:
         raise ValueError('invalid ftype')
 
     if zero_phase is None:
         warnings.warn(" Note: Decimate's zero_phase keyword argument will "
-                      "default to True in a future release.  Until then, "
-                      "decimate defaults to 'one-way filtering for backwards "
+                      "default to True in a future release. Until then, "
+                      "decimate defaults to one-way filtering for backwards "
                       "compatibility. Ideally, always set this argument "
                       "explicitly.", FutureWarning)
         zero_phase = False
 
-    if zero_phase and ftype == 'fir':
-        warnings.warn('zero_phase is not implemented for FIR downsampling, '
-                      'using zero_phase=False.')
-        zero_phase = False
-
-    if zero_phase:
-        y = filtfilt(system.num, system.den, x, axis=axis)
-    else:
-        y = lfilter(system.num, system.den, x, axis=axis)
+    if len(system.den) == 1:  # FIR case
+        if zero_phase:
+            y = resample_poly(x, 1, q, axis=axis, window=system.num)
+            return y
+        else:
+            # upfirdn is generally faster than lfilter by a factor equal to the
+            # downsampling factor, since it only calculates the needed outputs
+            n_out = int(np.ceil(x.shape[axis] / q))
+            y = upfirdn(system.num, x, up=1, down=q, axis=axis)
+            sl = [slice(None)] * y.ndim
+            sl[axis] = slice(None, n_out, None)
+            return y[sl]
+
+    else:  # IIR case
+        if zero_phase:
+            y = filtfilt(system.num, system.den, x, axis=axis)
+        else:
+            y = lfilter(system.num, system.den, x, axis=axis)
 
-    sl = [slice(None)] * y.ndim
-    sl[axis] = slice(None, None, q)
-    return y[sl]
+        sl = [slice(None)] * y.ndim
+        sl[axis] = slice(None, None, q)
+        return y[sl]
```

Comment:
Instead of having four different `return` paths, I usually prefer having one `return` at the end if possible, so each of these conditionals would just assign `y`. Thoughts?